srk/src/transitionSystem.ml

```diff
@@ -662,18 +684,168 @@ module Make
+    PHT.add pmap (entry, -1) PS.empty;
+    PHT.add qmap (entry, -1) PPS.empty;
+    let procedures = 
+      WG.fold_edges (fun (_, w, _) acc -> 
+        match w with 
+        | Call (x, y) -> 
+          begin match PHT.find_opt pmap (x, y) with 
+          | None -> 
+            PHT.add pmap (x, y) PS.empty 
+          | Some _ -> ()
+          end;
+          begin match PHT.find_opt qmap (x, y) with 
+          | None -> 
+            PHT.add qmap (x, y) PPS.empty 
+          | Some _ -> () 
+          end;
+          PS.add (x, y) acc
+        | _ -> acc 
+        ) tg (PS.add (entry, -1) PS.empty) in 
+    let rec dfs (proc: int * int) tg src (f: (vertex * vertex) -> vertex -> vertex * 'a label * vertex -> unit) (visited : ISet.t) = 
+      Printf.printf "dfs: visiting %d\n" src;
+      WG.fold_succ_e (fun (u, w, v) visited -> 
+        f proc src (u, w, v);
+        begin match ISet.find_opt v visited with 
+            | None -> ISet.union (dfs proc tg v f (ISet.add v visited)) visited 
+            | Some _ -> visited 
+            end) tg src visited in 
+      PS.iter (fun (x, y) -> 
+        Printf.printf "populating p/qmaps with dfs... %d %d\n" x y;
+        ignore @@ dfs (x, y) tg x (fun proc src (_, w, v) -> 
+          match w with 
+          | Call (x, y) -> 
+            let caller_set = PHT.find pmap proc in 
+            let callee_set = PHT.find qmap (x, y) in
+            PHT.add pmap proc (PS.add (x, y) caller_set);
+            PHT.add qmap (x, y) (PPS.add (proc, (src, v)) callee_set);
+          | _ -> () 
+          ) (ISet.add x ISet.empty)) procedures; (* populate pmap, qmap *)
+      let compute_sinks () = 
+        PS.fold (fun (x, y) acc -> (* compute sink locations to start inlining from *)
+          match (PHT.find_opt pmap (x, y), PHT.find_opt qmap (x, y)) with 
+          | Some callees, Some callers -> 
+            (* a procedure is considered for inlining if it is (1) a sink in the call graph (2) at least one function calls it.*)
+            if ((PS.cardinal callees) == 0) && ((PPS.cardinal callers) > 0) then PS.add (x, y) acc else 
+              begin 
+                Printf.printf "%d %d is not a sink; num callees = %d num callers = %d\n" x y (PS.cardinal callees) (PPS.cardinal callers);
+                acc
+              end 
+          | (_, _) -> failwith ""
+          ) procedures PS.empty in
+      let copy_subgraph tg src = 
+        let vertices = dfs (-1, -1) tg src (fun _ _ _ -> ()) (ISet.add src ISet.empty) in 
+        let to_map = Hashtbl.create 998 in 
+        let rtg = ref tg in 
+          ISet.iter (fun x -> 
+            greatest := !greatest + 1;
+            Hashtbl.add to_map x (!greatest);
+            begin match SrkUtil.Int.Map.find_opt x !assertions with (* if x in assertions, new vertex also in assertions *)
+              | Some assert_value -> 
+                assertions := SrkUtil.Int.Map.add !greatest assert_value !assertions 
```

Comment:
In my view, there are two possible types of clients
1. We have a set of assertions in the program, and we want to prove or disprove each separately.
2. We don't care to distinguish between asserts -- we want to prove all of them or refute one of them.
In case 1, we shouldn't duplicate asserts because, for instance, the client won't be able to make sense of an assert being both safe and unsafe (which can happen if the function containing the assert is duplicated for two different call sites).  In case 2, we also shouldn't duplicate asserts because we want to maintain the invariant that there's only one.

I suggest that we add a comment to inline that we must have the pre-condition that assert nodes have no outgoing edges, and inline should only copy non-assert nodes.